### PR TITLE
Move conquest system to world server

### PR DESCRIPTION
**Pull request by xkoredev**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

- Weekly tally / hourly updates now happen on world server
- World server communicates of map server updates via zmq
- Adds xi_world configuration and debug symbols for Mac OS
- Maps cache all data. Only world server uses DB.

## Steps to test these changes

Kill stuff as level 15+ outside. Check that influence points are changed in db. Check region map.
Run !updateconquest 0 for tally.
Run !updateconquest 2 for conquest update, after killing lots of mobs.

### CODE REVIEW

**Comment by xkoredev** [comment]
@zach2good

File: src/map/CMakeLists.txt [hunk 1]
```diff
@@ -137,9 +137,9 @@ set(SOURCES
 )
 
 if(UNIX)
-    set(resource "")
+    set(resource conquest_zmq.cpp conquest_zmq.h conquest_data.cpp conquest_data.h)
 else()
-    set(resource "${CMAKE_SOURCE_DIR}/res/mapserver.rc")
+    set(resource "${CMAKE_SOURCE_DIR}/res/mapserver.rc" conquest_zmq.cpp conquest_zmq.h conquest_data.cpp conquest_data.h)
```
Comment (zach2good) [inline src/map/CMakeLists.txt]: These don't go here, these go in the SOURCES section above - this is just for the rc  icon

File: src/map/conquest_data.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     * Loads the latest conquest data from DB.
+     * Should only ever be done on map initialization.
+     * World server should be updating conquest data periodically.
+     */
+    void load(std::unique_ptr<SqlConnection>& sql);
+
+    /**
+     * Gets the influence points for a given nation in a given region.
+     */
+    int32 getInfluence(REGION_TYPE region, uint8 nation);
```
Comment (zach2good) [inline src/map/conquest_data.h]: Can we start upgrading nation to its own enum like REGION_TYPE?

File: src/map/conquest_data.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Adds the given influence points to the given nation for the given
+     * region.
+     */
+    void addInfluencePoints(int points, uint8 nation, REGION_TYPE region);
+    /**
+     * Updates the influence points to match those given.
+     */
+    void updateInfluencePoints(std::vector<influence_t> influencePoints);
```
Comment (zach2good) [inline src/map/conquest_data.h]: If we're copying in influencePoints, presumably we're also not modifying it. So we could change this to be a const& and save ourselves a copy

File: src/map/conquest_data.h (hunk 2 above)
Comment (zach2good) [inline src/map/conquest_data.h]: Can you go through all of the methods you're adding and try to make them const, where possible (and take in larger data types as const&, where possible)

File: src/map/conquest_system.cpp [hunk 4]
```diff
@@ -34,90 +35,34 @@
 
 /************************************************************************
  *                                                                       *
- *  Реализация namespace conquest                                       *
+ *  Conquest namespace implementation                                    *
```
Comment (zach2good) [inline src/map/conquest_system.cpp]: 🔪 🇷🇺 

(I actually quite like Russia as a country, but I take great joy when these old comments get axed)

Comment (zach2good) [reply src/map/conquest_system.cpp]: Does this comment bring any value though? `namespace conquest` tells us its the start of the namespace...

File: src/map/conquest_zmq.cpp [hunk 5]
```diff
@@ -0,0 +1 @@
+#include "conquest_zmq.h"
```
Comment (zach2good) [inline src/map/conquest_zmq.cpp]: 👀

File: src/map/conquest_zmq.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#ifndef SERVER_CONQUEST_ZMQ_H
+#define SERVER_CONQUEST_ZMQ_H
+
+#include <zmq.hpp>
```
Comment (zach2good) [inline src/map/conquest_zmq.h]: Not needed?

File: src/map/message.cpp [hunk 7]
```diff
@@ -656,6 +658,95 @@ namespace message
...
+                            }
+
+                            conquest::conquestData->updateRegionControls(regionControls);
+                            return;
+                        }
+
+                        break;
+                    }
+                    default:
+                        ShowWarning("Message: unhandled regional event type %d", eventType);
```
Comment (zach2good) [inline src/map/message.cpp:746]: This needs to be in a {} block

File: src/map/utils/battleutils.cpp [hunk 8]
```diff
@@ -301,7 +301,7 @@ namespace battleutils
 
         if (ret != SQL_ERROR && sql->NumRows() != 0)
         {
-            for (uint32 x = 0; sql->NextRow() == SQL_SUCCESS; ++x)
+            while (sql->NextRow() == SQL_SUCCESS)
```
Comment (zach2good) [inline src/map/utils/battleutils.cpp:304]: 👍

File: src/map/utils/fishingcontest.cpp [hunk 9]
```diff
@@ -0,0 +1,832 @@
+﻿/*
```
Comment (zach2good) [inline src/map/utils/fishingcontest.cpp]: This shouldn't be part of this PR

File: src/world/CMakeLists.txt [hunk 10]
```diff
@@ -27,7 +27,7 @@ list(REMOVE_ITEM SOURCES
 if(UNIX)
     set(resource "")
 else()
-    set(resource "${CMAKE_SOURCE_DIR}/res/worldserver.rc")
+    set(resource "${CMAKE_SOURCE_DIR}/res/worldserver.rc" message_handler.h)
```
Comment (zach2good) [inline src/world/CMakeLists.txt]: This should go in the SOURCES block

File: src/world/conquest_system.cpp [hunk 11]
```diff
@@ -20,3 +20,342 @@ along with this program.  If not, see http://www.gnu.org/licenses/
...
+===========================================================================
+
+Copyright (c) 2010-2015 Darkstar Dev Teams
+
+                        This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+                                         (at your option) any later version.
+
+                                         This program is distributed in the hope that it will be useful,
```
Comment (zach2good) [inline src/world/conquest_system.cpp]: Formatting of this is all screwwy?

File: src/world/conquest_system.h [hunk 12]
```diff
@@ -20,9 +20,37 @@ along with this program.  If not, see http://www.gnu.org/licenses/
 */
 #pragma once
 
-class ConquestSystem
+#include "./common/sql.h"
```
Comment (zach2good) [inline src/world/conquest_system.h]: No need for the unix-style ./'s

File: src/world/conquest_system.cpp [hunk 13]
```diff
@@ -20,3 +20,342 @@ along with this program.  If not, see http://www.gnu.org/licenses/
...
+}
+
+bool ConquestSystem::handleMessage(MSGSERVTYPE     type,
+                                   zmq::message_t* extra,
+                                   zmq::message_t* packet,
+                                   zmq::message_t* from,
+                                   in_addr         from_addr,
+                                   uint16          from_port)
+{
+    if (type != MSGSERVTYPE::MSG_MAP2WORLD_REGIONAL_EVENT)
```
Comment (zach2good) [inline src/world/conquest_system.cpp]: I like the abstract way of handling handleMessage and the message handlers, but this is quite wasteful. 

It goes through each registered handler (we'll eventually have a few: the 4x regional event things, party system, some other stuff, etc.) and then checks to see if thats the right place. During registration we could be mapping the message type to the destination and then doing a lookup to dispatch the message, instead of this loop

File: src/world/message_server.cpp [hunk 14]
```diff
@@ -248,11 +297,14 @@ void message_server_listen(const bool& requestExit)
     }
 }
 
-void message_server_init(const bool& requestExit)
+void message_server_init(const std::shared_ptr<SqlConnection>&               sql,
```
Comment (zach2good) [inline src/world/message_server.cpp]: You've activated my trap card (and I think you messaged me about this).

The underlying SQL handle we use in SqlConnection is ABSOLUTELY NOT THREAD SAFE EVEN SLIGHTLY. If you access it from a different thread from where it was created it'll wipe it out silently and then the next access will crash the system. message_server lives on its own thread, so you can't pass a shared_ptr to a SqlConnection created on the main thread to it, it needs to have its own one (like I had set up before).

There's no harm in the message_server, the parent world server, and each message-handling-regional-event-thing having their own SqlConnections as part of their classes, they're very cheap and SQL Server is very performant

File: src/world/message_server.cpp [hunk 15]
```diff
@@ -248,11 +297,14 @@ void message_server_listen(const bool& requestExit)
...
+                         const std::vector<std::shared_ptr<MessageHandler>>& handlers,
```
Comment (zach2good) [inline src/world/message_server.cpp]: Just as a general design thing, I would probably pass in an array of weak_ptrs here and then canonicalize them when we need to use the underlying object

File: src/map/conquest_system.h [hunk 16]
```diff
@@ -49,9 +51,9 @@ class CCharEntity;
 
 namespace conquest
 {
-    void UpdateConquestSystem(); // Update conquest information in the DB
+    extern ConquestData* conquestData; // Cached data with influences / region controls
```
Comment (zach2good) [inline src/map/conquest_system.h]: Avoid using extern here, we can construct this in other ways

**Review by zach2good — COMMENTED** [review]
Just a quick driveby before I start my work day.

In places where there's a big fat sql global: don't be afraid of using it. In places where there isn't, you can/should be creating a SqlConnection for that class.

Make sure you're sticking to the style guide: <https://github.com/LandSandBoat/server/wiki/Development-Guide-Cpp#naming--misc>

File: src/world/conquest_system.cpp (hunk 11 above)
Comment (Xaver-DaRed) [reply src/world/conquest_system.cpp]: It's also uneeded I would think.
The LSB copyright is at the top of the file, no need to add back the old DSP one

File: src/map/conquest_system.h (hunk 16 above)
Comment (xkoredev) [reply src/map/conquest_system.h]: How would you go about constructing that data? Not used to working with namespaces

File: src/map/conquest_system.cpp (hunk 4 above)
Comment (xkoredev) [reply src/map/conquest_system.cpp]: Ya I was considering removing some of them. wasn't sure if we wanted to leave them for consistency, but in this case it makes sense

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: how would you register these? I was thinking of a map of tuple<type, subtype> but the syntax was kinda ugly

File: src/map/packets/conquest_map.cpp [hunk 17]
```diff
@@ -36,77 +38,38 @@ CConquestPacket::CConquestPacket(CCharEntity* PChar)
     this->setType(0x5E);
     this->setSize(0xB4);
 
-    // TODO: Periodically cache this information inside conquest_system so that it
-    //     : isn't so on-demand
-    const char* Query = "SELECT region_id, region_control, region_control_prev, \
-                         sandoria_influence, bastok_influence, windurst_influence, \
-                         beastmen_influence FROM conquest_system;";
+    auto  conquestData     = conquest::conquestData;
```
Comment (zach2good) [inline src/map/packets/conquest_map.cpp]: This is creating a copy

File: src/map/conquest_system.h (hunk 16 above)
Comment (zach2good) [reply src/map/conquest_system.h]: You need to look into how these work: https://en.cppreference.com/w/cpp/language/storage_duration

main.cpp:
```cpp
#include "conquest.h"

int main(int argc, char** argv)
{
    conquest::init();
    
    auto* conquestData = conquest::getConquestData();

    return 0;
}
```

conquest.h
```cpp
#pragma once

namespace conquest
{
    struct ConquestData
    {

    };

    void init();
    ConquestData* getConquestData();
}
```

conquest.cpp
```cpp
#include "conquest.h"

namespace conquest
{
    // You could let the data be default constructed here
    ConquestData conquestData;

    // Or:
    // ConquestData* conquestData;

    void init()
    {
        // You could construct it with args here
        // conquestData = new ConquestData(...);

        // * Make sure you delete it later, or use shared_ptr
    }

    ConquestData* getConquestData()
    {
        // You could also return raw ptr or a shared ptr here, or a ref
        return &conquestData;
    }
}
```


^ It depends entirely on how/where you want to store the data. I would recommend NOT putting it into static storage (stored and constructed in a namespace/global scope).

Comment (zach2good) [reply src/map/conquest_system.h]: ^ Note that the storage is defined in the compilation unit, not the header *

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (zach2good) [reply src/world/conquest_system.cpp]: You're quite close:

In this block:
```cpp
    // Setup message handlers
    std::vector<std::shared_ptr<MessageHandler>> handlers;
    handlers.push_back(conquestSystem);
    messageServer = std::make_unique<message_server_wrapper_t>(zmqSql, handlers, std::ref(m_RequestExit));
```

You've already created the conquest system, just create a map of the defining message type to a shared_ptr to the ConquestSystem. The map will be of pointers to MessageHandlers, since it's an interface that's all fine.

Comment (zach2good) [reply src/world/conquest_system.cpp]: Yeah, instead of this long chain of fallthroughs, you should be mapping things to function pointers

File: src/map/CMakeLists.txt (hunk 1 above)
Comment (xkoredev) [reply src/map/CMakeLists.txt]: My IDE put them there.. will move em out.

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: the part that i'd like to make prettier is mapping the type and subtype to a handler. Few options:
- using a tuple<type, subtype> as key for the handler map
- using a nested map<type, map<subtype, handler>>
- using a uint16 key and just use key = type | (subtype << 8)

I am leaning towards the latter

Comment (zach2good) [reply src/world/conquest_system.cpp]: You have 2x sets of keys, there's no need to combine them or do a single dispatch.

You do one lookup of destination key to shared ptr value of the relevant message handler in the message server, then forward the payload.

You do the next lookup of subject (message type) once you're in the correct handler, mapping the key to a function pointer of the function that'll handle the payload

Comment (zach2good) [reply src/world/conquest_system.cpp]: Type:  REGIONALMSGTYPE : uint8
Subtype: CONQUESTMSGTYPE : uint8

File: src/map/CMakeLists.txt (hunk 1 above)
Comment (xkoredev) [reply src/map/CMakeLists.txt]: Done

File: src/map/conquest_system.cpp (hunk 4 above)
Comment (xkoredev) [reply src/map/conquest_system.cpp]: Removed.

File: src/world/conquest_system.h (hunk 12 above)
Comment (xkoredev) [reply src/world/conquest_system.h]: Removed

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (zach2good) [reply src/world/conquest_system.cpp]:
```cpp
#include <vector>
#include <string>
#include <memory>
#include <unordered_map>
#include <functional>
#include <iostream>

// Defined globally
enum REGIONAL_EVENT_HANDLERS
{
    CONQUEST,
};

// Defined so only conquest systems in world/map can see/care about it
enum CONQUEST_MESSAGES
{
    DO_THING,
};

class MessageHandler
{
public:
    virtual ~MessageHandler() = default;
    virtual void handleMessage(std::vector<int> payload) = 0;
};

class ConquestSystem : public MessageHandler
{
public:
    ConquestSystem()
    {
        // Set up internal handler lookup
        internalHandlerMap[CONQUEST_MESSAGES::DO_THING] = std::bind(&ConquestSystem::handleDoThing, this, std::placeholders::_1);
    }

    void handleMessage(std::vector<int> payload)
    {
        // Lookup key, call member function
        auto key = (CONQUEST_MESSAGES)payload[1];
        internalHandlerMap[key](payload);
    }

    void handleDoThing(std::vector<int> payload)
    {
        std::cout << "Called handleDoThing from ConquestSystem\n";
    }

private:
    std::unordered_map<CONQUEST_MESSAGES, std::function<void(std::vector<int>)>> internalHandlerMap;
};

int main()
{
    // Demo Setup
    std::unordered_map<REGIONAL_EVENT_HANDLERS, std::shared_ptr<MessageHandler>> handlerMap;
    handlerMap[REGIONAL_EVENT_HANDLERS::CONQUEST] = std::make_shared<ConquestSystem>();

    // Message Server gets message
    std::vector<int> message{
        REGIONAL_EVENT_HANDLERS::CONQUEST,
        CONQUEST_MESSAGES::DO_THING,
        42 // payload
    };

    // Extract lookup key
    auto regionalEventHandlerKey = (REGIONAL_EVENT_HANDLERS)message[0];

    // Use key to lookup the handler, then give the message to the handler (a copy in this case)
    handlerMap[regionalEventHandlerKey]->handleMessage(message);

    return 0;
}
```

Comment (zach2good) [reply src/world/conquest_system.cpp]: I'd probably make `virtual void handleMessage(std::vector<int> payload) = 0;` take `std::vector<int>&&` so things had to be `std::move`'d into them

File: src/map/conquest_system.h (hunk 16 above)
Comment (zach2good) [reply src/map/conquest_system.h:54]: If each messageHandler ends up being on its own thread, then they'll have to communicate using locks or with a lockless queue, also

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (zach2good) [reply src/world/conquest_system.cpp]: If each messageHandler ends up being on its own thread, then they'll have to communicate using locks or with a lockless queue, also

File: src/world/conquest_system.cpp [hunk 18]
```diff
@@ -20,3 +20,342 @@ along with this program.  If not, see http://www.gnu.org/licenses/
...
+    // 3- Send tally end Msg
+    sendRegionControlsMsg(CONQUEST_WORLD2MAP_WEEKLY_UPDATE_END);
+}
+
+void ConquestSystem::updateHourlyConquest()
+{
+    sendInfluencesMsg(true);
+}
+
+void ConquestSystem::updateDailyTallyPoints()
```
Comment (cocosolos) [inline src/world/conquest_system.cpp]: Not sure if this is the best place for this since it has no relation to the conquest system.

File: src/map/conquest_data.h (hunk 2 above)
Comment (xkoredev) [reply src/map/conquest_data.h]: Done

File: src/map/conquest_data.h (hunk 3 above)
Comment (xkoredev) [reply src/map/conquest_data.h]: Done

File: src/map/conquest_data.h (hunk 2 above)
Comment (xkoredev) [reply src/map/conquest_data.h]: To be clear created the enum and used it in conquest, but wont refactor outside

Comment (xkoredev) [reply src/map/conquest_data.h]: Done

File: src/map/conquest_zmq.h (hunk 6 above)
Comment (xkoredev) [reply src/map/conquest_zmq.h]: Removed

File: src/map/conquest_zmq.cpp (hunk 5 above)
Comment (xkoredev) [reply src/map/conquest_zmq.cpp]: removed empty file

File: src/map/message.cpp (hunk 7 above)
Comment (xkoredev) [reply src/map/message.cpp:746]: Done. Curious, is this a codestyle thing (makes sense) or is the functionality different without it?

File: src/world/CMakeLists.txt (hunk 10 above)
Comment (xkoredev) [reply src/world/CMakeLists.txt]: Moved

File: src/map/utils/fishingcontest.cpp (hunk 9 above)
Comment (xkoredev) [reply src/map/utils/fishingcontest.cpp]: Removed. Must have been an artifact of my local state at the time in ASB

File: src/world/conquest_system.cpp (hunk 11 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: Thanks. Idk why I copy/pasta'd it there. Removed.

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: Since we don't know our requirements for all handlers yet, I was going to just keep the same arguments to the message parse method (`zmq::message_t* extra, zmq::message_t* packet, from_ip, from_port`). Would there be issues with that?

Comment (xkoredev) [reply src/world/conquest_system.cpp]: Actually packet seems more meant for the request / reply mechanism so I think I can do with just the data vector as you suggested, and from_ip / port

Comment (zach2good) [reply src/world/conquest_system.cpp]: Remember (so far, it might change in the future) that the idea is to not generate actual packet data for the client in the world server.

It's a little annoying and inconvenient that they use the same structures and mechanisms, but ultimately, world server doesn't have all the same information that map does - nor does it matter.

Comment (xkoredev) [reply src/world/conquest_system.cpp]: so if we send around an std::vector<uint8>, is this the way we would access properties?
```
        int32 points;
        std::memcpy(&points, payload.data(), sizeof(int32));
```

Any other options?

File: src/world/message_server.cpp (hunk 14 above)
Comment (xkoredev) [reply src/world/message_server.cpp]: I've switched to creating instances of sql connection as needed, or using the global when available.

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: I've changed this to use a vector<uint8> payload. Let me know if that makes sense.

File: src/world/message_server.cpp (hunk 15 above)
Comment (xkoredev) [reply src/world/message_server.cpp]: Gotcha. I am setting up the handlers in message server now so no need to do this anymore.

File: src/world/conquest_system.cpp (hunk 18 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: Moved it back. I was confused by the word tally. Thanks!

File: src/world/conquest_system.cpp (hunk 13 above)
Comment (zach2good) [reply src/world/conquest_system.cpp]: What do you mean by properties? For a given message type, you know what type of item is going in, vs what is coming out?
```
std::memcpy(&points, payload.data(), sizeof(int32));
```
Should work fine, `.data()` can be abused for storage of whatever you want - given that it's contiguous memory. If you want to store something more complex, it's defined as a vector of uint8's (bytes). So you can just construct the payload as sizeof(your_struct) and then write/read out the data from `.data()`

Comment (zach2good) [reply src/world/conquest_system.cpp]: If you _really really_ wanted to, you could also start using flatbuffers, but that feels overkill for this

File: src/map/conquest_system.h (hunk 16 above)
Comment (xkoredev) [reply src/map/conquest_system.h]: Done

File: src/map/packets/conquest_map.cpp (hunk 17 above)
Comment (xkoredev) [reply src/map/packets/conquest_map.cpp]: Using shared ptr now

**Comment by xkoredev** [comment]
@zach2good ammended comments. Next step will be adding the cache on the world server that is flushed every 2 min, as well as map server zone ip / port cache, tho the latter will require a world restart when maps restart if the settings change.
Should I make these changes in this PR? or follow up?

File: src/map/conquest_system.h [hunk 19]
```diff
@@ -60,8 +50,9 @@ namespace conquest
     uint8 GetInfluenceRanking(int32 san_inf, int32 bas_inf, int32 win_inf, int32 bst_inf);
     uint8 GetInfluenceRanking(int32 san_inf, int32 bas_inf, int32 win_inf);
 
-    void UpdateConquestGM(ConquestUpdate type); // Update conquest system by GM (modify in the DB and use @updateconquest)
-    void UpdateWeekConquest();                  // Update conquest system every sunday
+    void HandleInfluenceUpdate(const std::vector<influence_t>& influences, bool shouldUpdateZones); // Triggered periodically when world server sends updates of the latest influence data.
```
Comment (zach2good) [inline src/map/conquest_system.h]: We're an east const (EAST IS BEAST) family around here ;)

(apart from const char*, because what is life without nonsensical conventions)

File: src/map/conquest_data.cpp [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+#include "common/sql.h"
+#include "conquest_system.h"
+
+ConquestData::ConquestData()
+: regionControls(std::vector<region_control_t>(256))
+, influences(std::vector<influence_t>(256))
+{
+}
+
+void ConquestData::load(std::unique_ptr<SqlConnection>& sql)
```
Comment (zach2good) [inline src/map/conquest_data.cpp:14]: Given that this doesn't have any dependencies apart from the sql handle, I'd move all of this logic into the constructor and use the global sql handle with extern. I think its OK to lean on that design pattern for now, and at some point in the future I'll kill all the global extern usage of the sql handle etc.

Comment (zach2good) [reply src/map/conquest_data.cpp:14]: Also, given how little this data is, we could just lazy-load it just before its needed the first time?

File: src/world/world_server.cpp [hunk 21]
```diff
@@ -21,12 +21,13 @@ along with this program.  If not, see http://www.gnu.org/licenses/
...
 #include "common/application.h"
-#include "common/console_service.h"
 #include "common/logging.h"
+#include "common/vana_time.h"
 
 namespace
 {
-    std::unique_ptr<SqlConnection> sql;
+    std::unique_ptr<SqlConnection>  sql;
+    std::unique_ptr<ConquestSystem> conquest;
```
Comment (zach2good) [inline src/world/world_server.cpp]: There's already a std::shared_ptr<ConquestSystem> conquestSystem; in worldServer members?

File: src/world/world_server.cpp [hunk 22]
```diff
@@ -21,12 +21,13 @@ along with this program.  If not, see http://www.gnu.org/licenses/
...
 
 #include "common/application.h"
-#include "common/console_service.h"
 #include "common/logging.h"
+#include "common/vana_time.h"
 
 namespace
 {
-    std::unique_ptr<SqlConnection> sql;
+    std::unique_ptr<SqlConnection>  sql;
```
Comment (zach2good) [inline src/world/world_server.cpp:29]: World server itself isn't doing any SQL stuff, does it need its own sql handle?

File: src/world/message_server.cpp [hunk 23]
```diff
@@ -39,18 +40,59 @@ struct chat_message_t
 
 moodycamel::ConcurrentQueue<chat_message_t> outgoing_queue;
 
-std::unique_ptr<SqlConnection> zmqSql;
+std::unique_ptr<SqlConnection>                                       sql;
+std::unordered_map<REGIONALMSGTYPE, std::shared_ptr<MessageHandler>> regionalMsgHandlers;
+
+typedef std::function<bool(zmq::message_t* data,
```
Comment (zach2good) [inline src/world/message_server.cpp:46]: 👍

File: src/world/message_server.cpp [hunk 24]
```diff
@@ -163,6 +205,15 @@ void message_server_parse(MSGSERVTYPE type, zmq::message_t* extra, zmq::message_
             // no op
             break;
         }
+        case MSG_MAP2WORLD_REGIONAL_EVENT:
+        {
+            uint8*      data    = (uint8*)extra->data();
+            const uint8 subType = ref<uint8>(data, 0);
+
+            std::vector<uint8> bytes(data, data + extra->size());
+            regionalMsgHandlers[(REGIONALMSGTYPE)subType]->handleMessage(bytes, from_ip, from_port);
```
Comment (zach2good) [inline src/world/message_server.cpp]: 👍

Comment (zach2good) [reply src/world/message_server.cpp]: In the interest of safety, I'd look into not using operator[] and using .get() instead, and wrapping this in a try/catch block -> because world-server needs to be bulletproof. 

If an unvalidated subType comes in, operator[] access in the map will create a new shared_ptr, which we immediately call handleMessage on. I'm not sure if this would be a nullptr dereference or not, but best not to risk it

File: src/world/message_handler.h [hunk 25]
```diff
@@ -0,0 +1,18 @@
+#ifndef SERVER_MESSAGE_HANDLER_H
+#define SERVER_MESSAGE_HANDLER_H
+
+#include "common/mmo.h"
+#include <zmq.hpp>
+
+class MessageHandler
```
Comment (zach2good) [inline src/world/message_handler.h]: I'd probably call this IMessageHandler, since its a pure interface

File: src/world/conquest_system.cpp [hunk 26]
```diff
@@ -20,3 +20,282 @@ along with this program.  If not, see http://www.gnu.org/licenses/
...
+        const std::size_t offset             = headerLength + sizeof(size_t) + sizeof(region_control_t) * i;
+        ref<uint8>((uint8*)data, offset)     = regionControls[i].current;
+        ref<uint8>((uint8*)data, offset + 1) = regionControls[i].prev;
+    }
+
+    // 3- Create ZMQ Message and queue it
+    zmq::message_t dataMsg = zmq::message_t(dataLen);
+    memcpy(dataMsg.data(), data, dataLen);
+
+    if (ipp == 0)
```
Comment (zach2good) [inline src/world/conquest_system.cpp]: I wonder if this is dangerous or not, because I think we use 0 as a designator for "Zone is turned off" in a lot of places?

File: src/map/message.cpp (hunk 7 above)
Comment (zach2good) [reply src/map/message.cpp:746]: It functions because it just runs from the case statement downwards, but that doesn't align with what if and while do, so its cleaner and clearer to wrap things in {}'s, with the break either inside or outside - I don't know if we have a convention for that

**Review by zach2good — COMMENTED** [review]
Getting closer. I wonder if ConquestData could/should be a private class inside the conquest system file in the conquest namespace. It doesn't do much apart from data loading and retrieval, and those can all just be functions that exist in the conquest namespace?

File: src/world/message_server.cpp (hunk 24 above)
Comment (zach2good) [reply src/world/message_server.cpp]: [https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*](https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*)

Turns out this sort of access is undefined behaviour, lets avoid that :D

File: src/map/conquest_system.h (hunk 19 above)
Comment (xkoredev) [reply src/map/conquest_system.h]: Done

File: src/map/conquest_data.cpp (hunk 20 above)
Comment (xkoredev) [reply src/map/conquest_data.cpp:13]: I'll lazy load it

Comment (xkoredev) [reply src/map/conquest_data.cpp:13]: actually this is in map. which global should we rely on ? the map one?

Comment (xkoredev) [reply src/map/conquest_data.cpp:14]: Done. On construction

File: src/world/world_server.cpp (hunk 21 above)
Comment (xkoredev) [reply src/world/world_server.cpp]: We need to expose world in the namespace to then right?

Comment (xkoredev) [reply src/world/world_server.cpp]: This pointer is the one in world, assigned on construction, just like sql

File: src/world/world_server.cpp (hunk 22 above)
Comment (xkoredev) [reply src/world/world_server.cpp:29]: its used in this `UpdateDailyTallyPoints` method which I moved back outside per request (since it doesnt belong to conquest system)

File: src/world/message_server.cpp (hunk 24 above)
Comment (xkoredev) [reply src/world/message_server.cpp]: So I understand. Avoid using ->get() or avoid accessing without checking first?

File: src/world/message_handler.h (hunk 25 above)
Comment (xkoredev) [reply src/world/message_handler.h]: Done

File: src/world/conquest_system.cpp (hunk 26 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: Alternative would be create a different method, or maybe a boolean flag for broadcast?

**Comment by xkoredev** [comment]
> Getting closer. I wonder if ConquestData could/should be a private class inside the conquest system file in the conquest namespace. It doesn't do much apart from data loading and retrieval, and those can all just be functions that exist in the conquest namespace?

Ya I think it should. I just didn't know about that being a possibility as I have not worked with namespaces in the past. Will do.

**Comment by xkoredev** [comment]
> Contributor

I stacked overlown it (because im a namespace noob) and not sure that you can create a private namespace class:
https://stackoverflow.com/questions/38795042/how-make-a-class-private-inside-a-namespace-in-c#:~:text=You%20cannot.,restrict%20access%20using%20friend%20specifiers.

File: src/world/message_server.cpp (hunk 24 above)
Comment (zach2good) [reply src/world/message_server.cpp]:
```cpp
try
{
    auto& handler = regionalMsgHandlers.get((REGIONALMSGTYPE)subType); // or is this a shared ptr? I don't remember
    handler->handleMessage(...);
}
catch (the possible exception from get() throwing)
{
...
}
```

Comment (zach2good) [reply src/world/message_server.cpp]: The idea is to use get, because get will throw if the key isn't there. If you use operator[] and it isn't there, it'll create an entry (an empty shared ptr). Accessing members of the underlying class of an empty shared ptr is undefined behaviour (it might work, it might crash, we can't know)

File: src/world/conquest_system.cpp (hunk 26 above)
Comment (zach2good) [reply src/world/conquest_system.cpp]: If you're trying to use a magic number to send things to the world server, I'd probably go for 0xFFFF

**Comment by zach2good** [comment]
By private I mean it's defined only in the cpp file, so it isn't advertised to consumers of the related header.

```cpp
#include <vector>

// This is main.cpp, or any cpp file:
namespace conquest
{
    class Test
    {
        public:
            Test() = default;
    };
}

// conquest::Test won't be available from the conquest header, if thats also defining and advertising the conquest namespace

int main()
{
    conquest::Test t;   
    return 0;
}
```

File: src/world/world_server.cpp (hunk 21 above)
Comment (zach2good) [reply src/world/world_server.cpp]: An anonymous namespace is used to hide local objects/data within a given compilation unit - stopping them from being extern'd from elsewhere. You can _technically_ do either way, but both is messy and confusing. The Conquest System should be a unique ptr member of the class, it doesn't need to be hidden. The local SQL ptr makes more sense in the anonymous namespace because it is explicitly _not_ for sharing, since most systems on world server will have their own and everything is threaded

Comment (zach2good) [reply src/world/world_server.cpp]:
```cpp
WorldServer::WorldServer(int argc, char** argv)
: Application("world", argc, argv)
, conquestSystem()
, httpServer(std::make_unique<HTTPServer>())
, messageServer(std::make_unique<message_server_wrapper_t>(std::ref(m_RequestExit)))
{
    // Anonymous namespace preparation
    sql      = std::make_unique<SqlConnection>();
    conquest = std::make_unique<ConquestSystem>();
```

This means you have 2x conquest systems now. `, conquestSystem()` is constructing the WorldServer::conquestSystem member, and `conquest = std::make_unique<ConquestSystem>();` is constructing the anonymous namespace member conquest in this compilation unit.

**Comment by zach2good** [comment]
See here:
image

Even though conquest::Test is defined in the cpp, the definition isn't available to the header - unless you were to use a forward declaration of it, but that's not the point here

File: src/world/message_server.cpp (hunk 24 above)
Comment (xkoredev) [reply src/world/message_server.cpp]: gotcha thanks for explaining.

File: src/world/conquest_system.cpp (hunk 26 above)
Comment (xkoredev) [reply src/world/conquest_system.cpp]: 👍

File: src/world/world_server.cpp (hunk 21 above)
Comment (xkoredev) [reply src/world/world_server.cpp]: Oh right yes. I did that intentionally because conquestSystem is stateless and I did not know a better way to share the conquest system. However i'd prefer being able to use the world system instance from world_server.cpp, but how do we do that? (i.e for the tally tick logic)

File: src/world/message_server.cpp (hunk 24 above)
Comment (xkoredev) [reply src/world/message_server.cpp]: Done

**Comment by xkoredev** [comment]
> See here: image
> 
> Even though conquest::Test is defined in the cpp, the definition isn't available to the header - unless you were to use a forward declaration of it, but that's not the point here

Actually currently CCConquestPaquet accesses conquestData via the public method (conquest::GetConquestData()). Making conquestData private would require a lot more refactor

**Comment by zach2good** [comment]
https://godbolt.org/z/hdYrn117E

**Comment by xkoredev** [comment]
> 

Thanks for the handhold lol.. feel kinda bad. I misunderstood this as you wanting to avoid exposing `GetConquestData()`

**Comment by xkoredev** [comment]
> https://godbolt.org/z/hdYrn117E

The only downside I see here then is that conquest data has a few methods which will all be moved to conquest system, making the file a bit harder to parse.

**Comment by xkoredev** [comment]
> https://godbolt.org/z/hdYrn117E

How about accessing conquestData.x? This is what CConquestPacket() does, but I don't see a way of doing so without exposing this in the headers / API.

File: src/world/world_server.cpp (hunk 21 above)
Comment (xkoredev) [reply src/world/world_server.cpp]: Done. The same shared_ptr can be used for both

**Comment by xkoredev** [comment]
> 

As mentioned offline, we are keeping it public
